pkg/domain/infra/abi/farm.go

```diff
@@ -0,0 +1,119 @@
+	case 2:
+		os, arch = platform[0], platform[1]
+	case 3:
+		os, arch, variant = platform[0], platform[1], platform[2]
+	}
+	os, arch, variant = libimage.NormalizePlatform(os, arch, variant)
+	nativePlatform = os + "/" + arch
+	if variant != "" {
+		nativePlatform += ("/" + variant)
+	}
+	emulatedPlatforms = emulation.Registered()
+	return os, arch, variant, append([]string{}, nativePlatform), emulatedPlatforms, nil
+}
+
+// FarmNodeInspect returns information about the remote engines in the farm
+func (ir *ImageEngine) FarmNodeInspect(ctx context.Context) (*entities.FarmInspectReport, error) {
+	ir.platforms.Do(func() {
+		ir.os, ir.arch, ir.variant, ir.nativePlatforms, ir.emulatedPlatforms, ir.platformsErr = ir.fetchInfo(ctx)
+	})
+	return &entities.FarmInspectReport{NativePlatforms: ir.nativePlatforms,
+		EmulatedPlatforms: ir.emulatedPlatforms,
+		OS:                ir.os,
+		Arch:              ir.arch,
+		Variant:           ir.variant}, ir.platformsErr
+}
+
+// PullToFile pulls the image from the remote engine and saves it to a file,
+// returning a string-format reference which can be parsed by containers/image.
+func (ir *ImageEngine) PullToFile(ctx context.Context, options entities.PullToFileOptions) (reference string, err error) {
+	saveOptions := entities.ImageSaveOptions{
+		Format: options.SaveFormat,
+		Output: options.SaveFile,
+	}
+	if err := ir.Save(ctx, options.ImageID, nil, saveOptions); err != nil {
+		return "", fmt.Errorf("saving image %q: %w", options.ImageID, err)
+	}
+	return options.SaveFormat + ":" + options.SaveFile, nil
+}
+
+// PullToFile pulls the image from the remote engine and saves it to the local
+// engine passed in via options, returning a string-format reference which can
+// be parsed by containers/image.
+func (ir *ImageEngine) PullToLocal(ctx context.Context, options entities.PullToLocalOptions) (reference string, err error) {
+	destination := options.Destination
+
+	// already present at destination?
+	var br *entities.BoolReport
+	if destination == nil {
+		br, err = ir.Exists(ctx, options.ImageID)
+	} else {
+		br, err = destination.Exists(ctx, options.ImageID)
+	}
+	if err != nil {
+		return "", err
+	}
+	if br.Value {
+		return istorage.Transport.Name() + ":" + options.ImageID, nil
+	}
+
+	tempFile, err := os.CreateTemp("", "")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tempFile.Name())
+	defer tempFile.Close()
+
+	saveOptions := entities.ImageSaveOptions{
+		Format: options.SaveFormat,
+		Output: tempFile.Name(),
+	}
+	if err := ir.Save(ctx, options.ImageID, nil, saveOptions); err != nil {
+		return "", fmt.Errorf("saving image %q: %w", options.ImageID, err)
+	}
+
+	loadOptions := entities.ImageLoadOptions{
+		Input: tempFile.Name(),
+	}
+	if destination == nil {
+		_, err = ir.Load(ctx, loadOptions)
+	} else {
+		_, err = destination.Load(ctx, loadOptions)
+	}
+	if err != nil {
+		return "", err
+	}
+
+	return istorage.Transport.Name() + ":" + options.ImageID, nil
```

Comment:
I think the image ID may change.  Can use the name from image-load report instead?  If we want to return the image ID of the loaded image, we can inspect the image by the returned name.

Reply:
The image ID changes if there's a format conversion.  The `SaveFormat` value specifies an archive format which should avoid that.